File: java/src/main/java/algorithm/utils/errors/shapes/InvalidTriangleError.java
package algorithm.utils.errors.shapes;

import java.util.List;

import algorithm.utils.errors.ShapeError;

/**
 * <h1>InvalidTriangleError</h1>
 *
 * Exception raised when the provided side lengths cannot form a valid triangle.
 *
 * <h2>Example</h2>
 *
 * <pre>
 * <code>
 * import algorithm.utils.errors.shapes.InvalidTriangleError;
 *
 * var sideA = 3.0;
 * var sideB = 4.0;
 * var sideC = 10.0;
 * var err = new InvalidTriangleError(sideA, sideB, sideC);
 *
 * System.out.println(err);  // INVALID TRIANGLE
 * </code>
 * </pre>
 */
public final class InvalidTriangleError extends ShapeError {
    /**
     * Initializes a new instance of <code>InvalidTriangleError</code>.
     * <p>
     * This constructor accepts three side lengths as `float` values.
     *
     * @param sideA The length of the first side.
     * @param sideB The length of the second side.
     * @param sideC The length of the third side.
     */
    public InvalidTriangleError(double sideA, double sideB, double sideC) {
        super("Triangle", List.of(sideA, sideB, sideC), "Side lengths do not form a triangle");
    }
}

File: python/src/algorithm/utils/errors/shapes/invalid_triangle_error.py
from typing import final

from .. import ShapeError


@final
class InvalidTriangleError(ShapeError):
    """
    InvalidTriangleError
    ====================
    Exception raised when the provided side lengths cannot form a valid triangle.

    Example
    -------
    ```python
    from algorithm.utils.errors import InvalidTriangleError

    side_a = 3.0
    side_b = 4.0
    side_c = 10.0
    err = InvalidTriangleError(side_a, side_b, side_c)

    print(err)  # INVALID TRIANGLE
    ```

    Note
    ----
    This error is raised when the sum of the lengths of any two sides is less
    than or equal to the length of the third side.
    """

    def __init__(self, side_a: float, side_b: float, side_c: float):
        """
        Initializes a new instance of `InvalidTriangleError`.

        This constructor accepts three side lengths as `float` values.

        - **Parameters**
            - **side_a** The length of the first side.
            - **side_b** The length of the second side.
            - **side_c** The length of the third side.
        """
        super().__init__(
            "Triangle",
            [side_a, side_b, side_c],
            "Side lengths do not form a triangle",
        )
